src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -1305,6 +1305,12 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
                 {
                     bodyValue = await httpContext.Request.ReadFromJsonAsync(jsonTypeInfo);
                 }
+                catch (BadHttpRequestException ex) when (ex.StatusCode == StatusCodes.Status413RequestEntityTooLarge)
+                {
+                    Log.RequestBodyIOException(httpContext, ex);
+                    httpContext.Response.StatusCode = ex.StatusCode;
+                    return (null, false);
+                }
                 catch (IOException ex)
                 {
                     Log.RequestBodyIOException(httpContext, ex);
```

Comment:
IOException should still set a non-default status code like if it's not going to be propagated, otherwise there are probably similar ways to hit the same issue.